File: prioritarr/health.py
from __future__ import annotations

from datetime import datetime, timezone

from prioritarr.database import Database


def check_liveness(
    db: Database,
    max_heartbeat_age_seconds: int = 300,
) -> dict:
    """Check that the DB is accessible and the heartbeat is fresh.

    Returns ``{"status": "ok"}`` when healthy, or
    ``{"status": "unhealthy", "reason": "..."}`` on failure.
    """
    try:
        ts = db.get_heartbeat()
    except Exception as exc:
        return {"status": "unhealthy", "reason": f"db_error: {exc}"}

    if ts is None:
        return {"status": "unhealthy", "reason": "no_heartbeat"}

    try:
        last = datetime.fromisoformat(ts)
    except ValueError:
        return {"status": "unhealthy", "reason": "invalid_heartbeat_ts"}

    # Ensure tz-aware comparison
    if last.tzinfo is None:
        last = last.replace(tzinfo=timezone.utc)

    now = datetime.now(timezone.utc)
    age_seconds = (now - last).total_seconds()

    if age_seconds > max_heartbeat_age_seconds:
        return {
            "status": "unhealthy",
            "reason": f"heartbeat_stale: {age_seconds:.0f}s > {max_heartbeat_age_seconds}s",
        }

    return {"status": "ok"}


def check_readiness(
    db: Database,
    dependency_status: dict[str, str],
) -> dict:
    """Return a readiness payload including dependency statuses and last heartbeat.

    ``dependency_status`` is a mapping of dependency name → status string,
    e.g. ``{"sonarr": "ok", "qbit": "unreachable"}``.
    """
    last_heartbeat = db.get_heartbeat()

    return {
        "status": "ok" if all(v == "ok" for v in dependency_status.values()) else "degraded",
        "dependencies": dependency_status,
        "last_heartbeat": last_heartbeat,
    }

File: prioritarr-kotlin/app/src/main/kotlin/org/cquemin/prioritarr/health/Health.kt
package org.cquemin.prioritarr.health

import org.cquemin.prioritarr.database.Database
import org.cquemin.prioritarr.schemas.DependencyStatus
import org.cquemin.prioritarr.schemas.HealthOk
import org.cquemin.prioritarr.schemas.HealthUnhealthy
import org.cquemin.prioritarr.schemas.ReadyResponse
import java.time.Duration
import java.time.Instant
import java.time.OffsetDateTime

/** Return null when healthy, else the unhealthy body. Matches check_liveness in python. */
fun checkLiveness(db: Database, maxHeartbeatAgeSeconds: Long = 300): Pair<Boolean, Any> =
    try {
        val ts = db.getHeartbeat()
        if (ts == null) {
            false to HealthUnhealthy(reason = "no_heartbeat")
        } else {
            val last = try {
                OffsetDateTime.parse(ts).toInstant()
            } catch (e: Exception) {
                return false to HealthUnhealthy(reason = "invalid_heartbeat_ts")
            }
            val age = Duration.between(last, Instant.now()).seconds
            if (age > maxHeartbeatAgeSeconds) {
                false to HealthUnhealthy(reason = "heartbeat_stale: ${age}s > ${maxHeartbeatAgeSeconds}s")
            } else {
                true to HealthOk()
            }
        }
    } catch (e: Exception) {
        false to HealthUnhealthy(reason = "db_error: ${e.message}")
    }

fun checkReadiness(db: Database, dependencies: Map<String, DependencyStatus>): ReadyResponse {
    val allOk = dependencies.values.all { it == DependencyStatus.OK }
    return ReadyResponse(
        status = if (allOk) "ok" else "degraded",
        dependencies = dependencies,
        last_heartbeat = db.getHeartbeat(),
    )
}
